### PR TITLE
[SiliconFlow] Optimize geometric and geometric_ for Ascend and MThreads

### PR DESCRIPTION
**Pull request by winfan1314**
### PR Category

Operator / OP Test / Benchmark

### Type of Change

Bug Fix / Performance Optimization

### Description

Add dedicated Ascend and MThreads implementations of `geometric` and `geometric_`, with accuracy regression tests and backend benchmark coverage.

### Issue

N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Accuracy

| Backend | Environment | Result |
|---|---|---:|
| Ascend 910B1 | CANN 9.1.1, torch-npu 2.10.0, FlagTree 0.6.1+ascend3.5 | 57 passed |
| MThreads MTT S5000 | torch-musa 2.9.1, FlagTree 0.6.0+mthreads3.6 | 57 passed |

```bash
python -m pytest -q -s tests/test_geometric.py --ref cpu --tb=short
```

#### Benchmark methodology

- Core shapes: `(1073741824,)`, `(64, 64)`, `(4096, 4096)`, `(64, 512, 512)`, and `(1024, 1024, 1024)`.
- Probability: `p=0.5`.
- Dtypes: FP16, FP32, and BF16.
- Mode: `kernel`; 15 cases per operator and 30 cases per backend.
- MThreads baseline: native `torch.Tensor.geometric_` / `torch.ops.aten.geometric`. Speedup is `Torch latency / Gems latency`; the geometric mean is computed over the five shapes for each operator/dtype group.
- Ascend records Gems latency using the repository benchmark path. Native latency and speedup are N/A with `skip_native`, because torch-npu falls back to CPU for the native geometric operation.

```bash
python -m pytest -q -s benchmark/test_geometric.py \
  --level core --record json --output performance.json --tb=short
```

#### Benchmark summary

MThreads MTT S5000:

| Operator | Dtype | Cases | Minimum speedup | Geometric mean speedup |
|---|---|---:|---:|---:|
| `geometric_` | `torch.float16` | 5 | 1.0088x | 1.1129x |
| `geometric_` | `torch.float32` | 5 | 1.3590x | 1.4187x |
| `geometric_` | `torch.bfloat16` | 5 | 0.9829x | 1.0863x |
| `geometric` | `torch.float16` | 5 | 1.6258x | 2.0947x |
| `geometric` | `torch.float32` | 5 | 2.7767x | 3.0640x |
| `geometric` | `torch.bfloat16` | 5 | 1.5945x | 2.0667x |

Ascend 910B1 / CANN 9.1.1:

| Operator | Dtype | Cases | Gems latency range (ms) | Native speedup |
|---|---|---:|---:|---:|
| `geometric_` | `torch.float16` | 5 | 0.010270–188.487527 | N/A |
| `geometric_` | `torch.float32` | 5 | 0.010381–189.872467 | N/A |
| `geometric_` | `torch.bfloat16` | 5 | 0.010639–196.656320 | N/A |
| `geometric` | `torch.float16` | 5 | 0.010246–188.080553 | N/A |
| `geometric` | `torch.float32` | 5 | 0.010438–189.894409 | N/A |
| `geometric` | `torch.bfloat16` | 5 | 0.010614–196.659919 | N/A |

#### Raw test stdout

Captured test output is included below, including emitted warnings.

<details>
<summary>Ascend 910B1 CANN 9.1.1 accuracy stdout</summary>

```text
.........................................................
=============================== warnings summary ===============================
../../usr/local/python3.11.16/lib/python3.11/site-packages/torch_npu/npu/__init__.py:539
  /usr/local/python3.11.16/lib/python3.11/site-packages/torch_npu/npu/__init__.py:539: UserWarning: You can set the device capability via the environment variable TORCH_NPU_DEVICE_CAPABILITY. The return value of get_device_capability is only for compatibility with PyTorch and does not represent the actual capability of the NPU device.
    warnings.warn(f"You can set the device capability via the environment variable TORCH_NPU_DEVICE_CAPABILITY. {warning_str}")

src/flag_gems/utils/device_info.py:58
  /workspace/FlagGems/src/flag_gems/utils/device_info.py:58: UserWarning: [device_info] torch_device_fn.get_device_capability returned None for device_id=0, fallback to (0, 0).
    warnings.warn(

tests/test_geometric.py::test_geometric_[dtype0-shape0]
  /workspace/FlagGems/tests/test_geometric.py:34: UserWarning: Cannot create tensor with interal format while allow_internel_format=False, tensor will be created with base format. (Triggered internally at ../torch_npu/csrc/aten/common/TensorFactories.cpp:340.)
    ref_ones = utils.to_reference(torch.ones_like(x))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
57 passed, 3 warnings in 96.09s (0:01:36)
```
</details>

<details>
<summary>MThreads MTT S5000 accuracy stdout</summary>

```text
.........................................................
57 passed in 9.57s
```
</details>

<details>
<summary>Ascend 910B1 CANN 9.1.1 benchmark stdout</summary>

```text

Operator: geometric_  Performance Test (dtype=torch.float16, mode=kernel,level=core)
Native baseline: N/A (torch-npu geometric_ falls back to CPU)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS                    N/A          188.487527                 N/A          [torch.Size([1073741824]), 0.5]
SUCCESS                    N/A            0.010270                 N/A          [torch.Size([64, 64]), 0.5]
SUCCESS                    N/A            2.951023                 N/A          [torch.Size([4096, 4096]), 0.5]
SUCCESS                    N/A            2.948202                 N/A          [torch.Size([64, 512, 512]), 0.5]
SUCCESS                    N/A          188.072774                 N/A          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric_  Performance Test (dtype=torch.float32, mode=kernel,level=core)
Native baseline: N/A (torch-npu geometric_ falls back to CPU)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS                    N/A          189.872467                 N/A          [torch.Size([1073741824]), 0.5]
SUCCESS                    N/A            0.010381                 N/A          [torch.Size([64, 64]), 0.5]
SUCCESS                    N/A            2.978613                 N/A          [torch.Size([4096, 4096]), 0.5]
SUCCESS                    N/A            2.978600                 N/A          [torch.Size([64, 512, 512]), 0.5]
SUCCESS                    N/A          189.865586                 N/A          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric_  Performance Test (dtype=torch.bfloat16, mode=kernel,level=core)
Native baseline: N/A (torch-npu geometric_ falls back to CPU)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS                    N/A          196.643974                 N/A          [torch.Size([1073741824]), 0.5]
SUCCESS                    N/A            0.010639                 N/A          [torch.Size([64, 64]), 0.5]
SUCCESS                    N/A            3.091240                 N/A          [torch.Size([4096, 4096]), 0.5]
SUCCESS                    N/A            3.090851                 N/A          [torch.Size([64, 512, 512]), 0.5]
SUCCESS                    N/A          196.656320                 N/A          [torch.Size([1024, 1024, 1024]), 0.5]

.
Operator: geometric  Performance Test (dtype=torch.float16, mode=kernel,level=core)
Native baseline: N/A (torch-npu geometric falls back to CPU)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS                    N/A          188.080553                 N/A          [torch.Size([1073741824]), 0.5]
SUCCESS                    N/A            0.010246                 N/A          [torch.Size([64, 64]), 0.5]
SUCCESS                    N/A            2.948116                 N/A          [torch.Size([4096, 4096]), 0.5]
SUCCESS                    N/A            2.949659                 N/A          [torch.Size([64, 512, 512]), 0.5]
SUCCESS                    N/A          188.063455                 N/A          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric  Performance Test (dtype=torch.float32, mode=kernel,level=core)
Native baseline: N/A (torch-npu geometric falls back to CPU)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS                    N/A          189.894409                 N/A          [torch.Size([1073741824]), 0.5]
SUCCESS                    N/A            0.010438                 N/A          [torch.Size([64, 64]), 0.5]
SUCCESS                    N/A            2.979206                 N/A          [torch.Size([4096, 4096]), 0.5]
SUCCESS                    N/A            2.977997                 N/A          [torch.Size([64, 512, 512]), 0.5]
SUCCESS                    N/A          189.885882                 N/A          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric  Performance Test (dtype=torch.bfloat16, mode=kernel,level=core)
Native baseline: N/A (torch-npu geometric falls back to CPU)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS                    N/A          196.659919                 N/A          [torch.Size([1073741824]), 0.5]
SUCCESS                    N/A            0.010614                 N/A          [torch.Size([64, 64]), 0.5]
SUCCESS                    N/A            3.092134                 N/A          [torch.Size([4096, 4096]), 0.5]
SUCCESS                    N/A            3.090767                 N/A          [torch.Size([64, 512, 512]), 0.5]
SUCCESS                    N/A          196.636881                 N/A          [torch.Size([1024, 1024, 1024]), 0.5]

.
=============================== warnings summary ===============================
../../usr/local/python3.11.16/lib/python3.11/site-packages/torch_npu/npu/__init__.py:539
  /usr/local/python3.11.16/lib/python3.11/site-packages/torch_npu/npu/__init__.py:539: UserWarning: You can set the device capability via the environment variable TORCH_NPU_DEVICE_CAPABILITY. The return value of get_device_capability is only for compatibility with PyTorch and does not represent the actual capability of the NPU device.
    warnings.warn(f"You can set the device capability via the environment variable TORCH_NPU_DEVICE_CAPABILITY. {warning_str}")

src/flag_gems/utils/device_info.py:58
  /workspace/FlagGems/src/flag_gems/utils/device_info.py:58: UserWarning: [device_info] torch_device_fn.get_device_capability returned None for device_id=0, fallback to (0, 0).
    warnings.warn(

benchmark/test_geometric.py::test_geometric
  /workspace/FlagGems/src/flag_gems/runtime/backend/_ascend/ops/geometric.py:129: UserWarning: Cannot create tensor with interal format while allow_internel_format=False, tensor will be created with base format. (Triggered internally at ../torch_npu/csrc/aten/common/TensorFactories.cpp:340.)
    return _fill(torch.empty_like(self), p, generator)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
2 passed, 3 warnings in 132.49s (0:02:12)
```
</details>

<details>
<summary>MThreads MTT S5000 benchmark stdout</summary>

```text

Operator: geometric_  Performance Test (dtype=torch.float16, mode=kernel,level=core)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               5.302560            5.256080               1.009          [torch.Size([1073741824]), 0.5]
SUCCESS               0.004240            0.003160               1.342          [torch.Size([64, 64]), 0.5]
SUCCESS               0.094920            0.084920               1.118          [torch.Size([4096, 4096]), 0.5]
SUCCESS               0.094960            0.084920               1.118          [torch.Size([64, 512, 512]), 0.5]
SUCCESS               5.303700            5.255700               1.009          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric_  Performance Test (dtype=torch.float32, mode=kernel,level=core)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               6.766660            4.769940               1.419          [torch.Size([1073741824]), 0.5]
SUCCESS               0.004240            0.003120               1.359          [torch.Size([64, 64]), 0.5]
SUCCESS               0.111400            0.076800               1.451          [torch.Size([4096, 4096]), 0.5]
SUCCESS               0.111400            0.076840               1.450          [torch.Size([64, 512, 512]), 0.5]
SUCCESS               6.763860            4.771060               1.418          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric_  Performance Test (dtype=torch.bfloat16, mode=kernel,level=core)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               5.225120            5.314520               0.983          [torch.Size([1073741824]), 0.5]
SUCCESS               0.004240            0.003160               1.342          [torch.Size([64, 64]), 0.5]
SUCCESS               0.093680            0.086760               1.080          [torch.Size([4096, 4096]), 0.5]
SUCCESS               0.093680            0.086720               1.080          [torch.Size([64, 512, 512]), 0.5]
SUCCESS               5.223100            5.314080               0.983          [torch.Size([1024, 1024, 1024]), 0.5]

.
Operator: geometric  Performance Test (dtype=torch.float16, mode=kernel,level=core)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               8.547520            5.255900               1.626          [torch.Size([1073741824]), 0.5]
SUCCESS               0.011720            0.003160               3.709          [torch.Size([64, 64]), 0.5]
SUCCESS               0.172340            0.084920               2.029          [torch.Size([4096, 4096]), 0.5]
SUCCESS               0.172180            0.084960               2.027          [torch.Size([64, 512, 512]), 0.5]
SUCCESS               8.545320            5.256140               1.626          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric  Performance Test (dtype=torch.float32, mode=kernel,level=core)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS              13.247399            4.770900               2.777          [torch.Size([1073741824]), 0.5]
SUCCESS               0.011920            0.003080               3.870          [torch.Size([64, 64]), 0.5]
SUCCESS               0.230900            0.076840               3.005          [torch.Size([4096, 4096]), 0.5]
SUCCESS               0.230920            0.076840               3.005          [torch.Size([64, 512, 512]), 0.5]
SUCCESS              13.276159            4.770480               2.783          [torch.Size([1024, 1024, 1024]), 0.5]


Operator: geometric  Performance Test (dtype=torch.bfloat16, mode=kernel,level=core)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               8.475440            5.314520               1.595          [torch.Size([1073741824]), 0.5]
SUCCESS               0.012000            0.003160               3.797          [torch.Size([64, 64]), 0.5]
SUCCESS               0.171560            0.086680               1.979          [torch.Size([4096, 4096]), 0.5]
SUCCESS               0.171160            0.086760               1.973          [torch.Size([64, 512, 512]), 0.5]
SUCCESS               8.474240            5.314680               1.594          [torch.Size([1024, 1024, 1024]), 0.5]

.
2 passed in 48.22s
```
</details>

### CODE REVIEW

**Comment by winfan1314** [comment]
@modao1234 the Mthreads CI has an env. issue.

**Review by modao1234 — APPROVED** [review]
